File: FrontEnd/src/algorithms/dfs-algorithm.ts
import { BaseAgent } from "./base-agent";
import { normalize } from "@/utils/game-utils";

/**
 * DFS (Depth-First Search) Algorithm
 * Systematic exploration with frequency-based scoring
 */
export class DFSAgent extends BaseAgent {
  constructor(wordList: string[]) {
    super(wordList.map(w => normalize(w)));
  }

  protected chooseBestGuess(possibleWords: string[]): string {
    // If only one word left, return it
    if (possibleWords.length === 1) {
      return possibleWords[0];
    }

    // First guess: use optimal starter
    if (this.currentGuesses.length === 0) {
      const starters = this.getOptimalStarters();
      for (const starter of starters) {
        if (possibleWords.includes(starter)) {
          return starter;
        }
      }
    }

    // Find word with highest frequency score
    let bestWord = possibleWords[0];
    let bestScore = -1;

    // Limit search to avoid performance issues
    const searchSpace = possibleWords.slice(0, Math.min(100, possibleWords.length));

    for (const word of searchSpace) {
      const score = this.scoreWordFrequency(word, possibleWords);
      if (score > bestScore) {
        bestScore = score;
        bestWord = word;
      }
    }

    return bestWord;
  }
}

File: backend/agent/dfs_algorithm.py
"""
DFS (Depth-First Search) Algorithm

Systematic exploration with frequency-based scoring.
This algorithm uses letter and position frequency to choose the best guess.
"""

from typing import List
from .base_agent import BaseAgent


class DFSAgent(BaseAgent):
    """
    DFS Algorithm Agent
    
    Strategy: Systematic exploration with frequency-based scoring
    - Uses optimal starting words
    - Scores words based on letter frequency in remaining possibilities
    - Chooses word with highest frequency score
    """
    
    def choose_best_guess(self, possible_words: List[str]) -> str:
        """
        Choose the best guess using frequency-based scoring
        
        Args:
            possible_words: List of words that match all previous patterns
            
        Returns:
            The word with the highest frequency score
        """
        # If only one word left, return it
        if len(possible_words) == 1:
            return possible_words[0]
        
        # First guess: use optimal starter
        if len(self.current_guesses) == 0:
            starters = self.get_optimal_starters()
            for starter in starters:
                if starter in possible_words:
                    return starter
        
        # Find word with highest frequency score
        best_word = possible_words[0]
        best_score = -1.0
        
        # Limit search to avoid performance issues
        search_space = possible_words[:min(100, len(possible_words))]
        
        for word in search_space:
            score = self.score_word_frequency(word, possible_words)
            if score > best_score:
                best_score = score
                best_word = word
        
        return best_word
